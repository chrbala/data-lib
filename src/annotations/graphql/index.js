// @flow

import type { AllReducerType } from '../../shared/types';

export default (graphql: *) => 
	(graphqlOptions: {}) => 
		(reducers: AllReducerType) => new graphql.GraphQLObjectType({

		})
	;
;
