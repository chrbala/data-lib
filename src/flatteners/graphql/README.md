# Graphql

The Validator determines if the data provided is compliant with the schema. It will return null if there is no error. If there are errors, it will give an error object in the shape of the schema.

It's important to note that fields are **non-nullable** by default and must be made nullable by using maybe(). Types pulled from the store, however, are the base type.

This module is *stateful* in that calling .graphql() changes state - it doesn't simply return a value. This is part of the design so you can design your schema shape in a different file from your graphql options.

This API is a bit of a trial and is subject to change. I'm also interested in hearing use cases that would work well in the single-schema library.

## Setup
```javascript
import create from 'single-schema';
import GraphqlModule, { createStore } from 'single-schema/flatteners/graphql';

import * as graphql from 'graphql';

const store = createStore();
const flatteners = {
	graphql: GraphqlModule({store, graphql}),
};

const { combine, array, map, maybe } = create(flatteners);
```

## Leaves
Leaves are functions that return graphql types. Leaves are required to be defined as valid graphql types.

```javascript
import { GraphQLString } from 'graphql';

const string = {
	graphql: () => GraphQLString,
};
```

## Operators

### Combine

Adds a GraphQLObjectType or GraphQLInputObjectType to the store.

```javascript
const user = combine({
	username: string,
});

user.graphql('output', {
	name: 'Person',
	// Other graphql options go here.
	// Fields can be set, but the type is not directly settable here,
	// as it is defined in combine()
});
user.graphql('input', {
	name: 'PersonInput',
	// Other graphql input options go here.
	// Fields can be set, but the type is not directly settable here
	// as it is defined in combine()	
});

/*
returns the equivalent of: new GraphqlObjectType({
	name: 'Person',
	fields: {
		name: { type: new GraphQLNonNull(GraphQLString) },
	},
});
*/
store.get('Person');

/*
returns the equivalent of: new GraphqlInputObjectType({
	name: 'PersonInput',
	fields: {
		name: { type: new GraphQLNonNull(GraphQLString) },
	},
});
*/
store.get('PersonInput');
```

### Array

Coerces all of the elements of an array to its type.

```javascript
const user = combine({
	aliases: array(string),
});

user.graphql('output', {
	name: 'Person',
});
user.graphql('input', {
	name: 'PersonInput',
});

/*
returns the equivalent of: new GraphqlObjectType({
	name: 'Person',
	fields: {
		aliases: { type: new GraphQLNonNull(new GraphQLList(
			new GraphQLNonNull(GraphQLString)
		)) },
	},
});
*/
store.get('Person');

/*
returns the equivalent of: new GraphqlInputObjectType({
	name: 'PersonInput',
	fields: {
		aliases: { type: new GraphQLNonNull(new GraphQLList(
			new GraphQLNonNull(GraphQLString)
		)) },
	},
});
*/
store.get('PersonInput');
```

### Map

Map is not supported in Graphql.

### Maybe

Converts the type to a nullable type. 

```javascript
const user = combine({
	username: maybe(string),
});

user.graphql('output', {
	name: 'Person',
});
user.graphql('input', {
	name: 'PersonInput',
});

/*
returns the equivalent of: new GraphqlObjectType({
	name: 'Person',
	fields: {
		username: { type: GraphQLString },
	},
});
*/
store.get('Person');

/*
returns the equivalent of: new GraphqlInputObjectType({
	name: 'PersonInput',
	fields: {
		username: { type: GraphQLString },
	},
});
*/
store.get('PersonInput');
```

You can also create recursive data shapes with a thunk.

```javascript
const node = combine(() => ({
	value: string,
	next: maybe(node),
}));

node.graphql('output', {
	name: 'Node',
});
node.graphql('input', {
	name: 'NodeInput',
});

/*
returns the equivalent of: 
const node = new GraphqlObjectType({
	name: 'Node',
	fields: () => ({
		username: { type: node },
	}),
});
*/
store.get('Node');

/*
returns the equivalent of: 
const nodeInput = new GraphqlInputObjectType({
	name: 'PersonInput',
	fields: () => ({
		username: { type: nodeInput },
	}),
});
*/
store.get('PersonInput');
```

### And

And is not supported in Graphql.