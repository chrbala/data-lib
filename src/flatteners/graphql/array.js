// @flow

import wrap from './wrap';

export default wrap(({GraphQLList, GraphQLNonNull}) => data => 
	new GraphQLNonNull(new GraphQLList(data)))
;
