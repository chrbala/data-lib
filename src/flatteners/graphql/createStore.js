// @flow

export default () => {
	const thunks = {};
	const values = {};
	
	const set = (name: string, value: *) => thunks[name] = value;
	const get = (name: string) => {
		if (!values[name])
			if (!thunks[name])
				throw new Error(`${name} not yet registered!`);
			else
				values[name] = thunks[name]();
		return values[name];
	};
	const all = () => ({
		thunks,
		values,
	});
	
	return {
		set,
		get,
		all,
	};
};
