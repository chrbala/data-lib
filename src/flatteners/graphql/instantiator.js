// @flow

import { NAME, VALUE } from './types';
import type { 
	InputType, 
	OutputType, 
	ByValueType,
	StoreType,
	GetChildType,
	InitialConfigType,
	ConfigType,
} from './types';
import { normalizeInput } from './shared';
import { mapObj } from '../../util/micro';

const applyAll = ({getValue, wrappers}) => wrappers.reduce(
	(acc, next) => next(acc)
, getValue());

type NormalizeChildType = (arg: {
	store: StoreType, 
	child: OutputType, 
	graphql: *, 
	getChildName: GetChildType,
}) => ByValueType;
const normalizeChild: NormalizeChildType = ({
	store, 
	child, 
	graphql: {GraphQLNonNull}, 
	getChildName,
}) => {
	let getValue;

	if (child.type === VALUE)
		getValue = child.getValue;
	else if (child.type === NAME)
		getValue = () => store.get(getChildName(child.getName()));

	return {
		type: VALUE,
		getValue: () => new GraphQLNonNull(getValue()),
		wrappers: child.wrappers,
	};
};

const getType = (store, child, graphql) => 
	applyAll(normalizeChild(store, child, graphql));

export default ({store, variations, graphql}: InitialConfigType) => 
	(childNode: InputType) => 
		({fields: configFields = {}, ...config}: ConfigType) => {
			const normalized = normalizeInput(childNode);
			normalized.register(config.name);

			variations.forEach(({createName, build, getChildName}) => {
				const allConfig = {
					...config,
					fields: () => mapObj(normalized.getChildren(),
						(child, key) => ({
							...configFields[key],
							type: getType({
								store, 
								child: normalizeInput(child()), 
								graphql, 
								getChildName,
							}),
						})
					),
				};

				const name = createName(config.name);
				store.set(name, () => build({...allConfig, name}));
			});
		};
