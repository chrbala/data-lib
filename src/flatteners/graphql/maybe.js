// @flow

import wrap from './wrap';

export default wrap(({ getNullableType }) => data => getNullableType(data));
