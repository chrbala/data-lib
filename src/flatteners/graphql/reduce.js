// @flow

import { NAME } from './types';
import type { ByNameType, InitialConfigType } from './types';
import type { AllReducerType } from '../../shared/types';
import Instantiator from './instantiator';

export default (config: InitialConfigType) => (children: AllReducerType) => {
	let name;
	let hasBeenSet = false;

	const getName = () => {
		if (!hasBeenSet)
			throw new Error('Value has not been set');
		return name;
	};
	const register = (_name: string) => {
		if (hasBeenSet)
			throw new Error('Can only set register graphql objects once');
		hasBeenSet = true;

		name = _name;
	};
	const getChildren = () => children;

	const out: ByNameType = {
		type: NAME,
		getName,
		register,
		getChildren,
		wrappers: [],
	};

	return (...args: Array<*>) => {
		if (!args.length)
			return out;

		const [graphqlConfig] = args;
		Instantiator(config)(out)(graphqlConfig);
	};
};
