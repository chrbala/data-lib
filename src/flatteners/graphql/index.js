// @flow

import reduce from './reduce';
import array from './array';
import maybe from './maybe';

import type { InitialConfigType } from './types';

const GraphQLFlattener = (config: InitialConfigType) => ({
	reduce: reduce(config),
	array: array(config),
	maybe: maybe(config),
});

export default GraphQLFlattener;
