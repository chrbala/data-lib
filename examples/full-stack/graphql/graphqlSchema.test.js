// @flow
/* eslint-disable flowtype/no-weak-types */

import { graphql, printSchema } from 'graphql';

import schema from './graphqlSchema';
import * as Database from 'examples/full-stack/database';
import Loaders from './loaders';

it('Schema snapshot', () => {
	expect(printSchema(schema)).toMatchSnapshot();
});

const Context = () => {
	const database = Database.create();
	return Object.freeze({
		database,
		loaders: Loaders({database}),
	});
};

const throwErrors = res => res.errors 
	? Promise.reject(res.errors)
	: res
;

const insert = (name, query) => (input, context, fragment) => 
	graphql(schema, query(fragment), {}, context, {input})
		.then(throwErrors)
		.then((res: any) => res.data[name])
	;

const insertPerson = insert('insertPerson', fragment => `
	mutation($input:personInput!) {
	  insertPerson(input:$input) {
	  	${fragment}
	  }
	}
`);

const insertFamily = insert('insertFamily', fragment => `
	mutation($input:familyInput!) {
	  insertFamily(input:$input) {
	  	${fragment}
  	}
	}
`);

const get = (queryName, {query, context, params}) =>
	graphql(schema, query, {}, context, params)
		.then(throwErrors)
		.then((res: any) => 
			res.data[queryName]
		);

const getNode = (id, context, fragment) => get('node', {
	query: `
		query($id:String!) {
			node(id:$id) {
				${fragment}
			}
		}
	`, 
	context, 
	params: {id},
});

const getPersonAll = (context, fragment) => get('personAll', {
	query: `
		query {
			personAll {
				${fragment}
			}
		}
	`,
	context,
	params: {},
});

const getFamilyAll = (context, fragment) => get('familyAll', {
	query: `
		query {
			familyAll {
				${fragment}
			}
		}
	`,
	context,
	params: {},
});

it('Person insertion and retrieval', async () => {
	const NAME = 'NAME';
	const context = Context();

	const { name, id } = await insertPerson({name: NAME}, context, `
		... on person {
			id
			name
		}
	`);
	expect(name).toBe(NAME);

	const personNode = await getNode(id, context, `
		... on person {
			id
			name
		}
	`);
	
	const actual = personNode;
	const expected = {
		id,
		name: NAME,
	};

	expect(actual).toEqual(expected);
});

it('Family insertion and retrieval', async () => {
	const NAMES = ['ONE', 'TWO', 'THREE', 'FOUR', 'FIVE'];
	const context = Context();

	const people = await Promise.all(
		NAMES.map(name => insertPerson({name}, context, `...on person {
			id
		}`))
	).then(res => res.map(({id}) => ({id})));

	const familyInput = {
		adults: people.slice(0, 2),
		children: people.slice(2, 4),
	};

	const { 
		id, 
		...familyInsertionResult 
	} = await insertFamily(familyInput, context, `...on family {
		id
		adults {
			name
		}
		children {
			name
		}
	}`);
	expect(id).toBeTruthy();

	const expected = {
		adults: NAMES.slice(0, 2).map(name => ({name})),
		children: NAMES.slice(2, 4).map(name => ({name})),
	};

	expect(familyInsertionResult).toEqual(expected);

	const { 
		id: fetchedId, 
		...familyFetchResult 
	} = await getNode(id, context, `...on family {
		id
		adults {
			name
		}
		children {
			name
		}
	}`);
	expect(id).toBe(fetchedId);
	expect(familyFetchResult).toEqual(expected);
});

it('personAll works', async () => {
	const NAME = 'asdf';
	const context = Context();
		
	const { id } = await insertPerson({name: NAME}, context, `
		id
	`);
	expect(id).toBeTruthy();

	const allPeople = await getPersonAll(context, `
		id
		name
	`);
	const insertedPerson = allPeople.find(({id: _id}) => id == _id);
	expect(insertedPerson.name).toBe(NAME);
});

it('familyAll works', async () => {
	const context = Context();
		
	const family = {
		adults: [],
		children: [],
	};
	const { id } = await insertFamily(family, context, `
		id
		adults {
			name
		}
		children {
			name
		}
	`);
	expect(id).toBeTruthy();

	const allFamilies = await getFamilyAll(context, `
		id
		adults {
			name
		}
		children {
			name
		}
	`);
	const insertedFamily = allFamilies.find(({id: _id}) => id == _id);
	expect(insertedFamily.adults).toEqual([]);
	expect(insertedFamily.children).toEqual([]);
});

it('Bad person ID is rejected before it is persisted', async () => {
	const context = Context();

	const familyInput = {
		adults: [{id: 'bogus'}],
		children: [],
	};

	const fragment = `
		id
	`;
	const before = (await getFamilyAll(context, fragment)).length;

	try {
		await insertFamily(familyInput, context, `
			id
		`);
	} catch (e) {}

	const after = (await getFamilyAll(context, fragment)).length;
	expect(typeof before).toBe('number');
	expect(before).toBe(after);
});
