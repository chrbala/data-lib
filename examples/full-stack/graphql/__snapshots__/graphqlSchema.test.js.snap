exports[`test Schema snapshot 1`] = `
"schema {
  query: query
  mutation: mutation
}

type family implements node {
  id: String!
  adults: [person!]!
  children: [person!]!
}

input familyInput {
  adults: [pointer!]!
  children: [pointer!]!
}

input familyMutation {
  clientMutationId: String
  family: familyInput!
}

type familyPayload {
  clientMutationId: String!
  family: family!
}

type mutation {
  insertPerson(input: personMutation!): personPayload!
  insertFamily(input: familyMutation!): familyPayload!
}

interface node {
  id: String!
}

type person implements node {
  id: String!
  name: String!
}

input personInput {
  name: String!
}

input personMutation {
  clientMutationId: String
  person: personInput!
}

type personPayload {
  clientMutationId: String!
  person: person!
}

input pointer {
  id: String!
}

type query {
  node(id: String!): node!
  viewer: viewer!
}

type viewer {
  personAll: [person!]!
  familyAll: [family!]!
}
"
`;
